### PR TITLE
gh-104635: Eliminate redundant STORE_FAST instructions in the compiler

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-104635

```
import pyperf

runner = pyperf.Runner()
runner.timeit(name="bench dead_store",
              stmt="""
_, _, _, a = 1, 2, 3, 4
""")


Mean +- std dev: [base] 10.8 ns +- 0.2 ns -> [super] 10.1 ns +- 0.1 ns: 1.07x faster
```

### CODE REVIEW

File: Lib/test/test_peepholer.py [hunk 1]
```diff
@@ -1071,7 +1071,17 @@ def test_no_unsafe_static_swap(self):
             ('POP_TOP', 0, 4),
             ('RETURN_VALUE', 5)
         ]
-        self.cfg_optimization_test(insts, insts, consts=list(range(3)), nlocals=1)
+        expected_insts = [
+            ('LOAD_CONST', 0, 1),
+            ('LOAD_CONST', 1, 2),
+            ('LOAD_CONST', 2, 3),
+            ('SWAP', 3, 4),
```
Comment (carljm) [inline Lib/test/test_peepholer.py]: I was surprised that this `SWAP` is still here, since allowing `apply_static_swaps` to eliminate it is the main point of this optimization. But on second thought it makes sense; `apply_static_swaps` reads instructions forward from the `SWAP` instruction that triggers it, and that will happen before we reach the `STORE_FAST` instruction that we will replace with `POP_TOP`. So `apply_static_swaps` isn't currently able to benefit from replacing the `STORE_FAST` with `POP_TOP`.

Short of making `optimize_basic_block` multi-pass, I think to address this we'd need to effectively duplicate the `STORE_FAST` -> `POP_TOP` optimization inside `apply_static_swaps`. Which isn't that bad; it's not a lot of code. That could also be a separate PR, though.

File: Python/flowgraph.c [hunk 2]
```diff
@@ -1515,6 +1515,14 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                     */
                 }
                 break;
+            case STORE_FAST:
+                if (opcode == nextop &&
+                    oparg == bb->b_instr[i+1].i_oparg &&
+                    bb->b_instr[i].i_loc.lineno == bb->b_instr[i+1].i_loc.lineno) {
```
Comment (carljm) [inline Python/flowgraph.c:1521]: we should have a test verifying the optimization doesn't occur when linenos are different

File: Python/instrumentation.c [hunk 3]
```diff
@@ -1493,6 +1493,7 @@ static const uint8_t super_instructions[256] = {
     [STORE_FAST__LOAD_FAST] = 1,
     [STORE_FAST__STORE_FAST] = 1,
     [LOAD_CONST__LOAD_FAST] = 1,
+    [POP_TOP__STORE_FAST] = 1,
```
Comment (carljm) [inline Python/instrumentation.c:1497]: what about `POP_TOP__POP_TOP` here?

**Review by carljm — COMMENTED** [review]
This seems worth doing if the reported 1% win from pyperformance is real. We can also do a bit better here by allowing this optimization to help `apply_static_swaps`.

I'm not sure if there are reasons to avoid new super-instructions? Would like feedback from @markshannon on that.

File: Lib/test/test_peepholer.py (hunk 1 above)
Comment (corona10) [reply Lib/test/test_peepholer.py]: How about adding apply_static_swap as the last phase of optimize_basic_block?
From the single responsibility principle, IIUC, it will be more convenient to implement. But it will cost more optimization time, but I am not sure that it will be acceptable. 

Maybe @iritkatriel can comment on this point.


````patch

--- a/Python/flowgraph.c
+++ b/Python/flowgraph.c
@@ -1528,10 +1528,6 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                     INSTR_SET_OP0(inst, NOP);
                     break;
                 }
-                if (swaptimize(bb, &i) < 0) {
-                    goto error;
-                }
-                apply_static_swaps(bb, i);
                 break;
             case KW_NAMES:
                 break;
@@ -1546,6 +1542,16 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                 assert (!HAS_CONST(inst->i_opcode));
         }
     }
+
+    for (int i = 0; i < bb->b_iused; i++) {
+        cfg_instr *inst = &bb->b_instr[i];
+        if (inst->i_opcode == SWAP) {
+            if (swaptimize(bb, &i) < 0) {
+                    goto error;
+            }
+            apply_static_swaps(bb, i);
+        }
+    }
     return SUCCESS;
 error:
     return ERROR;
````


## AS-IS
````

>>> import dis
>>> def f(x, y):
...     a, a = x, y
...     return a
...
>>> dis.dis(f)
  1           0 RESUME                   0

  2           2 LOAD_FAST                0 (x)
              4 LOAD_FAST                1 (y)
              6 SWAP                     2
              8 POP_TOP
             10 STORE_FAST               2 (a)

  3          12 LOAD_FAST                2 (a)
             14 RETURN_VALUE
````

## TO-BE
````
>>> import dis
>>> def f(x, y):
...     a, a = x, y
...     return a
...
>>> dis.dis(f)
  1           0 RESUME                   0

  2           2 LOAD_FAST                0 (x)
              4 LOAD_FAST                1 (y)
              6 STORE_FAST               2 (a)
              8 POP_TOP

  3          10 LOAD_FAST                2 (a)
             12 RETURN_VALUE
````

Comment (iritkatriel) [reply Lib/test/test_peepholer.py]: Optimization time is not currently a concern AFAIK.

File: Python/flowgraph.c (hunk 2 above)
Comment (corona10) [reply Python/flowgraph.c:1521]: I will add it!

File: Lib/test/test_peepholer.py [hunk 4]
```diff
@@ -1071,7 +1071,64 @@ def test_no_unsafe_static_swap(self):
...
+            ('RETURN_VALUE', 5)
+        ]
+        self.cfg_optimization_test(insts, expected_insts, consts=list(range(3)), nlocals=1)
+
+    def test_dead_store_elimination_in_same_lineno(self):
+        insts = [
+            ('LOAD_CONST', 0, 1),
+            ('LOAD_CONST', 1, 2),
+            ('LOAD_CONST', 2, 3),
+            ('SWAP', 3, 4),
```
Comment (carljm) [inline Lib/test/test_peepholer.py]: I think the SWAP is kind of irrelevant to this test?

File: Lib/test/test_peepholer.py [hunk 5]
```diff
@@ -1071,7 +1071,64 @@ def test_no_unsafe_static_swap(self):
...
+            ('RETURN_VALUE', 5)
+        ]
+        self.cfg_optimization_test(insts, expected_insts, consts=list(range(3)), nlocals=1)
+
+    def test_no_dead_store_elimination_in_different_lineno(self):
+        insts = [
+            ('LOAD_CONST', 0, 1),
+            ('LOAD_CONST', 1, 2),
+            ('LOAD_CONST', 2, 3),
+            ('SWAP', 3, 4),
```
Comment (carljm) [inline Lib/test/test_peepholer.py]: here also the `SWAP` seems irrelevant to the purpose of the test?

File: Misc/NEWS.d/next/Core and Builtins/2023-05-31-16-53-17.gh-issue-104635.iwf0d8.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+The basic dead store elimination is adopted in the basic block optimization
+process. Patch by Dong-hee Na and Carl Meyer.
```
Comment (carljm) [inline Misc/NEWS.d/next/Core and Builtins/2023-05-31-16-53-17.gh-issue-104635.iwf0d8.rst]:
```suggestion
Eliminate redundant :opcode:`STORE_FAST` instructions in the compiler. Patch by Dong-hee Na and Carl Meyer.
```

File: Lib/test/test_peepholer.py (hunk 1 above)
Comment (carljm) [reply Lib/test/test_peepholer.py]: I don't see any other optimizations in `optimize_basic_block` that would depend on having the `SWAP` optimizations run first, so the idea of moving the `SWAP` optimization to a separate second loop over the block seems reasonable enough.

**Review by carljm — COMMENTED** [review]
This looks reasonable to me! Seems ok to do the swap thing either in this PR or separately. Interested in any feedback from @iritkatriel / @markshannon / @brandtbucher.

**Comment by carljm** [comment]
We might also want to expand this so it can look more than one instruction ahead for another write to the same location, skipping over an allowlist of instructions that we know don't read that same local, don't execute arbitrary code, and can't raise an execution. E.g. `LOAD_FAST` of a different local, `POP_TOP`, `NOP`, probably others.

**Comment by corona10** [comment]
@mdboom Hi, Can we run the pyperformance benchmark for this PR?

**Comment by corona10** [comment]
> We might also want to expand this so it can look more than one instruction ahead for another write to the same location, skipping over an allowlist of instructions that we know don't read that same local, don't execute arbitrary code, and can't raise an exception.

Yeah, great idea.
Currently, we can not optimize the following case.
````python
_, _, a, _ = 1, 2, 3, 4
```` 

But if we expand the optimization, we can do it.
Do we have to expand the optimization in this PR or handle it on a separate PR?

File: Lib/test/test_peepholer.py [hunk 7]
```diff
@@ -1071,7 +1071,57 @@ def test_no_unsafe_static_swap(self):
...
+            ('STORE_FAST', 1, 4),
+            ('STORE_FAST', 1, 4),
+            ('STORE_FAST', 1, 4),
+            ('RETURN_VALUE', 5)
+        ]
+        expected_insts = [
+            ('LOAD_CONST', 0, 1),
+            ('LOAD_CONST', 1, 2),
+            ('NOP', 0, 3),
+            ('POP_TOP', 0, 4),
```
Comment (carljm) [inline Lib/test/test_peepholer.py:1098]: This is a cool benefit I didn't realize we'd also get! `LOAD_*/POP_TOP` pairs can be eliminated entirely.

**Review by carljm — APPROVED** [review]
Looks good to me, presuming pyperformance comes back looking neutral or better.

> Do we have to expand the optimization in this PR or handle it on a separate PR?

Either way seems fine to me.

**Comment by carljm** [comment]
> Do you have any ideas or suggestions?

I think we should wait for #105230 to land, given the motivations outlined in #105229. Then we should re-measure the performance impact of this. I think it would be good to do a stats run of pyperformance with this change and see how often `POP_TOP` is followed by another `POP_TOP` or by `STORE_FAST`, in order to help inform whether we should add `POP_TOP_POP_TOP` and `POP_TOP_STORE_FAST` superinstructions (in the compiler).

File: Lib/test/test_peepholer.py (hunk 7 above)
Comment (brandtbucher) [reply Lib/test/test_peepholer.py:1098]: Ideally this would just become `LOAD_CONST NOP NOP STORE_FAST` though, right?

Comment (carljm) [reply Lib/test/test_peepholer.py:1098]: Yes; I think that's an orthogonal improvement to the `LOAD/POP` elimination optimization, allowing it to ignore intervening `NOP`. Or maybe it has to do rather with making it multi-pass? Or both? I haven't checked.

Comment (corona10) [reply Lib/test/test_peepholer.py:1098]: Maybe we can run the loop of optimization until there are no opcode changes.
(I didn't test yet)

**Comment by corona10** [comment]
>  I think it would be good to do a stats run of pyperformance with this change and see how often POP_TOP is followed by another POP_TOP or by STORE_FAST, 

When I ran the pyperformance, they showed the very low ratio of executions.
* POP_TOP__POP_TOP: 4,846,500 counts == 0.0 %
* POP_TOP__STORE_FAST: 19,918,740 counts  == 0.0%
https://gist.github.com/corona10/810255492523f2fa802396ca1deac73f

**Comment by corona10** [comment]
Replaced by https://github.com/python/cpython/pull/105320
